### ISSUE TITLE
Abstract shape computed incorrectly for `lax.scan`

### ISSUE THREAD

**Issue by sharadmv** [google/jax#598]
The abstract eval for scan ([`_scan_abstract_eval`](https://github.com/google/jax/blob/master/jax/lax.py#L1081-L1082)) returns `aval_out`, but in the case of scan, the `aval_out` should have an additional dimension corresponding to the dimension the scan is looping over (`scan` aggregates over this axis whereas `while_loop` and `fori_loop` don't). 

This crops up when using `jit` on `lax.scan` functions that use the output of `lax.scan` (if you just return the output of `lax.scan` it works fine). Here's how to reproduce an error:

```python
import jax.numpy as np
from jax import lax, jit

def f1(bs):
    result = lax.scan(lambda a, _: a, np.ones(5), bs)
    return result
def f2(bs):
    result = lax.scan(lambda a, _: a, np.ones(5), bs)
    return result.transpose((1, 0)).transpose((1, 0))
fast_f1 = jit(f1)
fast_f2 = jit(f2)

f1(np.ones(10)) #works
f2(np.ones(10)) #works
fast_f1(np.ones(10)) #works
fast_f2(np.ones(10)) #errors 
```
In theory all of these should be the same, but because we call `transpose` and the abstract result only has one dimension, it errors when it is traced.

Any suggestions on the best way to fix? I came up with a hacky fix where I change
```python
def _scan_abstract_eval(a, bs, consts, aval_out, jaxpr):
  return maybe_tracer_tuple_to_abstract_tuple(aval_out)
```
to
```python
def _scan_abstract_eval(a, bs, consts, aval_out, jaxpr):
  aval_out = core.AbstractTuple(ShapedArray(shape=(bs[0].shape[0],) + a.shape, dtype=a.dtype) for a in aval_out)
  return maybe_tracer_tuple_to_abstract_tuple(aval_out)
```

**Comment by mattjj** [comment]
Thanks for raising this!

Actually we should probably delete the `lax.scan` on master. The real one is coming in the differentiable-scan branch. Still a work-in-progress, but we hope it will be ready in the next week or two.

Unfortunately that means for now there really is no scan...

**Comment by mattjj** [comment]
I made a PR that will "fix" this issue in that it'll delete our temporary scan implementation. It's better this way because the new scan may have a different interface, and we don't want user code to start depending on this old interface.

I'm really glad you're digging deeply into JAX :D

**Comment by sharadmv** [comment]
Thanks for the update!
